### PR TITLE
Support >2G model export - alternative implementation | torchlib(feat)

### PR DESCRIPTION
**Pull request by justinchuby**
Support >2G model export - alternative implementation where we build initializers ourselves. This is a follow up of #1003. We now build the TensorProto ourselves directly from PyTorch tensors. This circumvents torchscript _export_onnx's limitation of 2G protobuf serialization and additional serialization, because we are now keeping everything in memory.

2G model is now no longer a special case because we add initializers in a separate step.

### TODO

- [ ] Test to_model_proto with initlizers for all torch data types.

### CODE REVIEW

**Comment by justinchuby** [comment]
@BowenBao Speed seems to be reasonably good. There's errors: `Unsuported type proto value case.`

**Comment by BowenBao** [comment]
Adding back infer_shapes and check_model made tests pass locally.

**Comment by justinchuby** [comment]
Let me test again

**Comment by justinchuby** [comment]
> Adding back infer_shapes and check_model made tests pass locally.

Nice, thanks! I updated this PR.

**Review by BowenBao — APPROVED** [review]
🚀

**Comment by justinchuby** [comment]
Needs validation with real models

File: onnxscript/function_libs/torch_lib/graph_building.py [hunk 1]
```diff
@@ -281,12 +277,38 @@ def eval_function(  # type: ignore[override]
...
+    for name, tensor in initializers.items():
+        tensor = tensor.detach().cpu().contiguous()
+        # Take the raw data directly from the tensor to avoid the overhead of
+        # data manipulation in onnx.helper.make_tensor
+        raw_data = bytes(
+            (ctypes.c_ubyte * tensor.element_size() * tensor.numel()).from_address(
+                tensor.data_ptr()
+            )
+        )
+        tensor_proto = onnx.helper.make_tensor(
```
Comment (wschin) [inline onnxscript/function_libs/torch_lib/graph_building.py:295]: consider optionally making it `external_data` so you never keep the model in memory.

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/graph_building.py:295]: I think that's what we wanted (return a single model proto, so the interface is uniform)? If we didn't need the initializers we can set include_initializers to False. The exporter is responsible for deciding what to do with the model (save to external data, serialize etc.)

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/graph_building.py:295]: This way to_model_proto does not have side effects.

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/graph_building.py:295]: Are you concerned about the case when we don't have enough memory so we don't want to duplicate the data? I wonder how we want to design the export api for this case?

Comment (wschin) [reply onnxscript/function_libs/torch_lib/graph_building.py:295]: Yes. I meant. TensorProto can be stored as schema (name, shape, element_type) plus a path to a file containing its content -- it's a valid in-memory TensorProto. If we optionally enable it, we can make this part of code scalable to huge scale models. I haven't gone through all parts end-to-end, so this idea might be local instead global necessary.

Comment (wschin) [reply onnxscript/function_libs/torch_lib/graph_building.py:295]: There are multiple fields you can use under TensorProto to store tensor content; `float_data`, `raw_data`, and `external_data`. I feel we should let use to optionally use `external_data`.

Ref: https://github.com/onnx/onnx/blob/ec38b7c4cf3656b9115a68f525017a6e69aa8773/onnx/onnx-ml.proto#L608

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/graph_building.py:295]: I see! That makes sense. Let me think more on this

File: onnxscript/function_libs/torch_lib/graph_building.py [hunk 2]
```diff
@@ -378,12 +374,38 @@ def eval_function(  # type: ignore[override]
         return self._graph.add_function_call(function, inputs, attributes)
 
 
+def _add_initializers(
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_lib/graph_building.py:377]: fyi @ kunal-vaishnavi creating onnx tensorproto from torch tensor dataptr

**Pull request closed without merging**